sw/device/silicon_creator/lib/drivers/flash_ctrl.c

```diff
@@ -172,59 +240,71 @@ void flash_ctrl_status_get(flash_ctrl_status_t *status) {
       bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_INIT_WIP_BIT);
 }
 
-rom_error_t flash_ctrl_read(uint32_t addr, uint32_t word_count,
-                            flash_ctrl_partition_t partition, uint32_t *data) {
+rom_error_t flash_ctrl_data_read(uint32_t addr, uint32_t word_count,
+                                 uint32_t *data) {
   RETURN_IF_ERROR(transaction_start((transaction_params_t){
       .addr = addr,
       .op_type = FLASH_CTRL_CONTROL_OP_VALUE_READ,
-      .partition = partition,
+      .partition = kFlashCtrlPartitionData,
       .word_count = word_count,
       // Does not apply to read transactions.
       .erase_type = kFlashCtrlEraseTypePage,
   }));
   fifo_read(word_count, data);
-  return wait_for_done();
+  return wait_for_done(kErrorFlashCtrlDataRead);
 }
 
-rom_error_t flash_ctrl_prog(uint32_t addr, uint32_t word_count,
-                            flash_ctrl_partition_t partition,
-                            const uint32_t *data) {
-  enum {
-    kWindowWordCount =
-        FLASH_CTRL_PARAM_REG_BUS_PGM_RES_BYTES / sizeof(uint32_t),
-  };
-
-  // Find the number of words that can be written in the first window.
-  uint32_t window_word_count =
-      kWindowWordCount - ((addr / sizeof(uint32_t)) % kWindowWordCount);
-  while (word_count > 0) {
-    // Program operations can't cross window boundaries.
-    window_word_count =
-        word_count < window_word_count ? word_count : window_word_count;
-
-    RETURN_IF_ERROR(transaction_start((transaction_params_t){
-        .addr = addr,
-        .op_type = FLASH_CTRL_CONTROL_OP_VALUE_PROG,
-        .partition = partition,
-        .word_count = window_word_count,
-        // Does not apply to program transactions.
-        .erase_type = kFlashCtrlEraseTypePage,
-    }));
+rom_error_t flash_ctrl_info_read(flash_ctrl_info_page_t info_page,
+                                 uint32_t offset, uint32_t word_count,
+                                 uint32_t *data) {
+  const uint32_t addr = info_page_addr(info_page) + offset;
```

Comment:
Are the permission controls going to be added as separate functions? I go back and forth, but feel like it probably makes sense to keep them decoupled.

Reply:
Yes, the plan is to add a function like `flash_ctrl_info_mp_set(info_page, access)` and set required permissions in `boot_data`. For now, the functest sets required permissions at a broader scope.